### PR TITLE
Replace myalgo-connect with @perawallet/connect

### PR DESCRIPTION
**Pull request by mrcointreau**
PeraWallet connect integration

### CODE REVIEW

File: assets/arc-0031/client/.eslintrc [hunk 1]
```diff
@@ -1,4 +1,45 @@
 {
-  "root": true,
-  "extends": ["@nuxtjs/eslint-config-typescript", "plugin:prettier/recommended"]
+  "extends": [
```
Comment (deanstef) [inline assets/arc-0031/client/.eslintrc:2]: I don't understand this file. Leftover?

File: assets/arc-0031/client/middleware/guard.global.ts [hunk 2]
```diff
@@ -1,14 +1,12 @@
...
-  if (process.server) {
-    const authStore = useAuthStore()
-    const { loadSession } = authStore
-    loadSession()
-    if (!authStore.session && to.path !== '/signin') {
+export default defineNuxtRouteMiddleware((to) => {
+  if (import.meta.server) {
+    const session = useCookie<Session>('session')
+    if (!session.value && to.path !== '/signin') {
       return '/signin'
```
Comment (deanstef) [inline assets/arc-0031/client/middleware/guard.global.ts:7]: Just curious: why not use the `navigateTo()` method as shown [here](https://nuxt.com/docs/guide/directory-structure/middleware#usage)?

File: assets/arc-0031/client/utils/hooks/useAuth.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import { encode, decode } from '@msgpack/msgpack'
+
+import { useNotifications } from './useNotifications'
+import { usePeraWallet } from './usePeraWallet'
+
+export interface AuthMessage {
+  /** The domain name of the Verifier */
+  domain: string
+  /** Algorand account to authenticate with */
+  authAcc: string
```
Comment (deanstef) [inline assets/arc-0031/client/utils/hooks/useAuth.ts:12]: this should be a 32byte string

File: assets/arc-0031/client/utils/hooks/useAuth.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  /** Optional, metadata */
+  meta?: string
+}
+
+export interface Session {
+  authAcc: string
+  accessToken: string
+}
+
+export const useAuth = () => {
```
Comment (deanstef) [inline assets/arc-0031/client/utils/hooks/useAuth.ts:26]: Could you explain me what the `= () => {...` does?

File: assets/arc-0031/server/routes/api/v1.ts [hunk 5]
```diff
@@ -1,15 +1,29 @@
...
+import type { Request, Response } from 'express'
+import express from 'express'
 import { v4 as uuidv4 } from 'uuid'
-import users from '../../mock/users.json'
-import { type AuthMessage } from '../../types'
+
+import { env } from '../../env'
+import users from '../../mock/users.db.json' assert { type: 'json' }
+
+interface AuthMessage {
```
Comment (deanstef) [inline assets/arc-0031/server/routes/api/v1.ts:10]: This is duplicated across client and server. What about having a generic model?

**Review by deanstef — COMMENTED** [review]
Thank you for putting this together. It looks very nice. I have added a few minor comments.

File: assets/arc-0031/client/.eslintrc (hunk 1 above)
Comment (mrcointreau) [reply assets/arc-0031/client/.eslintrc:2]: This is the [eslint](https://eslint.org/) configuration file. When you run the linting script (via `npm run lint`) this is the configuration that is used for the linting.

File: assets/arc-0031/client/middleware/guard.global.ts (hunk 2 above)
Comment (mrcointreau) [reply assets/arc-0031/client/middleware/guard.global.ts:7]: It seems like returning the route path or using `navigateTo` with default parameters do the same thing: return the given path specifying a 302 (Found) http status code.

File: assets/arc-0031/client/.eslintrc (hunk 1 above)
Comment (deanstef) [reply assets/arc-0031/client/.eslintrc:2]: Ok, thanks! Do we need to run it manually after doing some edits?

File: assets/arc-0031/client/utils/hooks/useAuth.ts (hunk 3 above)
Comment (mrcointreau) [reply assets/arc-0031/client/utils/hooks/useAuth.ts:12]: In TypeScript, you can't directly specify a type that represents a string of a specific length. TypeScript's type system doesn't have a built-in way to enforce a specific string length at compile time. I can add a server side check via `algosdk.isValidAddress` in order to verify that `request.body.authAcc` is a valid Algorand Address.

File: assets/arc-0031/client/middleware/guard.global.ts (hunk 2 above)
Comment (deanstef) [reply assets/arc-0031/client/middleware/guard.global.ts:7]: Ok, thanks!

File: assets/arc-0031/client/utils/hooks/useAuth.ts (hunk 3 above)
Comment (deanstef) [reply assets/arc-0031/client/utils/hooks/useAuth.ts:12]: I see - thanks for the clarification. Yes, I think we should check if the address is valid.

File: assets/arc-0031/client/utils/hooks/useAuth.ts (hunk 4 above)
Comment (mrcointreau) [reply assets/arc-0031/client/utils/hooks/useAuth.ts:26]: The `const myFunction = () => {...} `notation is used to define an arrow function, while function `myFunction() {...}` is used to define a function in the traditional way. There are a few key differences between these two types of functions:

1. `this` keyword: In a traditional function, the `this` keyword represents the object that called the function. In an arrow function, `this` is lexically bound, meaning it retains the value of `this` from its enclosing context.

2. Arguments object: Traditional functions have an `arguments` object, which is an array-like object that contains all arguments passed to the function. Arrow functions do not have an `arguments` object. Instead, they have rest parameters.

3. Constructor: Traditional functions can be used as constructors to create new objects with the `new` keyword. Arrow functions cannot be used as constructors and will throw an error if used with `new`.

4. Implicit return: Arrow functions can have an implicit return if the function body is a single expression. This means you can omit the `return` keyword and the curly braces `{}`. Traditional functions always need a `return` statement to return a value (unless the function returns `undefined`).

Comment (deanstef) [reply assets/arc-0031/client/utils/hooks/useAuth.ts:26]: helpful, thank you!

File: assets/arc-0031/server/routes/api/v1.ts (hunk 5 above)
Comment (mrcointreau) [reply assets/arc-0031/server/routes/api/v1.ts:10]: I acknowledge that maintaining identical interfaces in two separate locations is not ideal for several reasons, including potential inconsistencies and the extra overhead in ensuring they remain synchronized.

To address this, the idea of introducing a shared folder at the root level containing common interfaces like AuthMessage was considered. This approach would indeed resolve the issue of duplication and ensure consistency between our client and server implementations. However, there are some significant implications to this approach, especially concerning our Docker setup:

 1. Docker Build Context and Configuration: If we include a shared folder, we would need to modify the Dockerfiles and docker-compose.yml to change the build context to encompass the entire project, including the shared folder. This change might lead to increased build times and larger Docker image sizes, as Docker would need to process more data.

 2. Project Complexity: While setting up a monorepo using tools like Turbopack or Nx could efficiently manage this shared structure, it introduces a new level of complexity. For our current project scale, particularly considering this change is primarily for a single interface, the added complexity might not be justifiable.

Given these considerations, I propose we continue with the current structure for the time being. While it does mean some duplication, this approach allows us to maintain our current, straightforward Docker configuration and avoid the potential overhead of a more complex project structure.

File: assets/arc-0031/client/.eslintrc (hunk 1 above)
Comment (mrcointreau) [reply assets/arc-0031/client/.eslintrc:2]: Yes, you can also add it to the pre-commit hook

File: assets/arc-0031/client/utils/hooks/useAuth.ts (hunk 3 above)
Comment (mrcointreau) [reply assets/arc-0031/client/utils/hooks/useAuth.ts:12]: Addressed [here](https://github.com/deanstef/ARCs/pull/2/commits/128b89f8a6bad3eaf9c854cd3a5684ebb52b5331)

File: assets/arc-0031/server/routes/api/v1.ts (hunk 5 above)
Comment (deanstef) [reply assets/arc-0031/server/routes/api/v1.ts:10]: Agreed -- let's keep it simple for now.

**Review by deanstef — APPROVED** [review]

**Pull request merged**